src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java

```diff
@@ -99,24 +100,31 @@ public AttestationTokenRetriever(Vertx vertx,
 
     private void attestationExpirationCheck(long timerId) {
         // This check is to avoid the attest() function takes longer than 60sec and get called again from this method while attesting.
-        if (this.isAttesting) {
+        if (!this.isAttesting.compareAndSet(false, true)) {
             LOGGER.warn("In the process of attesting. Skip re-attest.");
-        } else {
+            return;
+        }
+
+        try {
             Instant currentTime = clock.now();
             Instant tenMinutesBeforeExpire = attestationTokenExpiresAt.minusSeconds(600);
+            if (!currentTime.isAfter(tenMinutesBeforeExpire)) {
+                return;
+            }
+
+            LOGGER.info("Attestation token is 10 mins from the expiry timestamp {}. Re-attest...", attestationTokenExpiresAt);
 
-            if (currentTime.isAfter(tenMinutesBeforeExpire)) {
-                LOGGER.info("Attestation token is 10 mins from the expiry timestamp {}. Re-attest...", attestationTokenExpiresAt);
-                try {
-                    this.isAttesting = true;
-                    attest();
-                } catch (AttestationTokenRetrieverException | IOException e) {
-                    notifyResponseStatusWatcher(401);
-                    LOGGER.info("Re-attest failed: ", e);
-                } finally {
-                    this.isAttesting = false;
-                }
+            if (!attestationProvider.isReady()) {
```

Comment:
Maybe we should check this in the attest function as well and throw an exception if it's not ready

Other consumers of the retriever will call attest directly, it shouldn't proceed if the provider is not ready

Reply:
If we also check isReady in the attest function, then isReady check will be executed twice.

I am thinking if we must check isReady before each attest, probably we should not add a new isReady method but throw a NotReady exception by getAttestationRequest method. Basically the option 2 in https://atlassian.thetradedesk.com/jira/browse/UID2-1632

Reply:
I don't see any issue with isReady being called twice

But I don't mind if we throw an exception instead. In the attestationExpirationCheck we can catch and log, allowing it to be checked again a minute later. Up to you

Reply:
GCP isReady involves file opening which is not light. But considering attestation calls are not frequent, calling twice is fine.

Let me check isReady in attest also.

Reply:
Updated, thanks